lib/Backends/DeviceManager.cpp

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Backends/DeviceManager.h"
+
+using namespace glow;
+
+DeviceManager::DeviceManager(std::unique_ptr<Backend> backend)
+    : backend_(std::move(backend)), workThread_([this] { runLoop(); }) {}
+
+DeviceManager::~DeviceManager() {
+  stop(true); // will join workThread_
+}
+
+void DeviceManager::init() {}
+
+void DeviceManager::addNetwork(DeviceNetworkID moduleID,
+                               std::unique_ptr<Module> module,
+                               ReadyCB callback) {
+  enqueueOperation(
+      {NETWORK_ADD, moduleID, std::move(module), std::move(callback)});
+}
+
+void DeviceManager::evictNetwork(DeviceNetworkID moduleID) {
+  enqueueOperation({NETWORK_EVICT, moduleID});
+}
+
+void DeviceManager::runFunction(DeviceNetworkID moduleID,
+                                llvm::StringRef functionName,
+                                std::unique_ptr<Context> ctx,
+                                ResultCB callback) {
+  enqueueOperation({FUNCTION_EXECUTE, moduleID, std::move(functionName),
+                    std::move(ctx), std::move(callback)});
+}
+
+void DeviceManager::stop(bool block) {
+  stop_ = true;
+  if (block && workThread_.joinable()) {
+    workThread_.join();
+  }
+}
+
+void DeviceManager::enqueueOperation(DeviceManagerOperation op) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  operationQueue_.push_back(std::move(op));
+  lock.unlock();
+  wakeup_.notify_one();
+}
+
+void DeviceManager::runLoop() {
+  while (!stop_) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    while (!operationQueue_.empty()) {
+      DeviceManagerOperation op = std::move(operationQueue_.front());
+      operationQueue_.pop_front();
+      lock.unlock(); // let pushes to the queue occur while processing
+
+      switch (op.type) {
+      case NETWORK_ADD:
+        initNetwork(std::move(op));
+        break;
+      case NETWORK_EVICT:
+        evictNetwork(std::move(op));
+        break;
+      case FUNCTION_EXECUTE:
+        executeFunction(std::move(op));
+        break;
+      }
+
+      lock.lock();
+    }
+
+    // Wait for more work.
+    if (!operationQueue_.empty()) {
```

Comment:
Should this be `if(operationQueue_.empty())`? Also does the `if` need to be here since in order to be down here the while loop would have had to exited already?